### PR TITLE
Analyzer: Fix assert in tryReplaceAndEqualsChainsWithConstant

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp [hunk 1]
```diff
@@ -323,8 +325,21 @@ class LogicalExpressionOptimizerVisitor : public InDepthQueryTreeVisitorWithCont
...
-            node = std::move(and_operands[0]);
+
+            auto operand_type = and_operands[0]->getResultType();
+            auto function_type = function_node.getResultType();
+            if (operand_type->lowCardinality())
+            {
+                /// Result of equality operator can be low cardinality, while AND always returns UInt8.
+                /// In that case we replace `(lc = 1) AND (lc = 1)` with `(lc = 1) AS UInt8`
+                assert(!function_type->isNullable() && function_type->equals(*removeLowCardinality(operand_type)));
+                node = createCastFunction(std::move(and_operands[0]), function_type, getContext());
```
Comment (vdimir) [inline src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp:336]: I'm not sure if we need add CAST or re-resolve subtree

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit cc654d44f21cf54b6d5a78977ee51fce8bc9a2e8 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57139/cc654d44f21cf54b6d5a78977ee51fce8bc9a2e8/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6968688412">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57139/cc654d44f21cf54b6d5a78977ee51fce8bc9a2e8/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp [hunk 2]
```diff
@@ -323,8 +325,21 @@ class LogicalExpressionOptimizerVisitor : public InDepthQueryTreeVisitorWithCont
             /// Because we reduce the number of operands here by eliminating the same equality checks,
             /// the only situation we can end up here is we had AND check where all the equality checks are the same so we know the type is UInt8.
             /// Otherwise, we will have > 1 operands and we don't have to do anything.
-            assert(!function_node.getResultType()->isNullable() && and_operands[0]->getResultType()->equals(*function_node.getResultType()));
-            node = std::move(and_operands[0]);
+
+            auto operand_type = and_operands[0]->getResultType();
+            auto function_type = function_node.getResultType();
+            if (operand_type->lowCardinality())
```
Comment (KochetovNicolai) [inline src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp:331]: I suppose we can always add CAST if `function_type != operand_type`

**Review by KochetovNicolai — APPROVED** [review]

**Comment by vdimir** [comment]
I checked fuzzer, it's another issue in the same pass, I added commit to fix it too.

**Pull request merged**